Sources/NIOHTTP1/HTTPDecoder.swift

```diff
@@ -437,6 +469,40 @@ public class HTTPDecoder<HTTPMessageT>: ByteToMessageDecoder, AnyHTTPDecoder {
             ctx.close(promise: nil)
         }
     }
+
+    private func readEOF(ctx: ChannelHandlerContext) {
+        guard !self.errored else {
+            // We're in readEOF because we hit an error and closed the connection.
+            // No need to do this dance again, just return.
+            return
+        }
+
+        // EOF is semantic in HTTP, so we need to tell the parser that we saw it.
+        // We need to be a bit careful: the parser can call out to us, but should only
+        // ever call on_message_complete, which won't do a read. As a result, it *should*
+        // be totally safe to call this with a null data pointer. Just to make sure, though,
+        // let's store nil in the base address.
+        self.state.baseAddress = nil
+        _ = c_nio_http_parser_execute(&self.parser, &self.settings, nil, 0)
+
+        // No check to state.currentError becuase, if we hit it before, we already threw that
+        // error. This never calls any of the callbacks that set that field anyway. Instead we
+        // just check if the errno is set and throw.
+        let errno = parser.http_errno
+        if errno != 0 {
+            self.errored = true
+            ctx.fireErrorCaught(HTTPParserError.httpError(fromCHTTPParserErrno: http_errno(rawValue: errno))!)
+            return
+        }
+
+        // If we got here, no error was hit, but we may have a pending callout. Let's fire it if we did.
+        if self.pendingInOut.count > 0 {
+            self.channelReadComplete(ctx: ctx)
+        }
+
+        // We don't need the cumulation buffer, if we're holding it.
+        self.cumulationBuffer = nil
```

Comment:
@Lukasa should we set this to nil before we call `channelReadComplete` to reduce the reference count and so minimize the chances of CoW ?

Reply:
Sure, in fact we can set it to `nil` very early in this function without danger.